src/functionals.jl

```diff
@@ -30,5 +37,9 @@ end
 # end
 
 function kldivergence(P::UnivariateDistribution, Q::UnivariateDistribution)
-    expectation(P, x -> let p = pdf(P,x); (p > 0)*log(p/pdf(Q,x)) end)
+    expectation(P, x -> let p = pdf(P, x); (p > 0) * log(p / pdf(Q, x)) end)
+end
+
+function kldivergence(P::MultivariateDistribution, Q::MultivariateDistribution; kwargs...)
+    expectation(P, x -> let p = pdf(P, x); (p > 0) * log(p / pdf(Q, x)) end; kwargs...)
```

Comment:
Can we change both to use `logpdf` instead? The samples have to have non-zero probability but `pdf` could still be zero (e.g. far away from the mean of a normal distribution); I think this is more important in general than potential cancellations due to the subtraction:
```suggestion
    function kldivergence_f(x)
        logpdf_P = logpdf(P, x)
        return (logpdf_P > oftype(logpdf_P, -Inf)) * (logpdf_P - logpdf(Q, x))
    end
    return expectation(P, kldivergence_f; kwargs...)
end
```

Reply:
Somehow that does not sound very AD-friendly. Why not using `isinf` ?

Reply:
`isinf` would catch `+Inf` as well but we only want to check for `-Inf` here. It's not clear to me though why it would be less AD-friendly than `p > 0`?

Reply:
> isinf would catch +Inf 

That's true!

>why it would be less AD-friendly than p > 0?

I am just thinking that `oftype` could mess up some of the AD mechanisms, `Dual`s for example?

Reply:
Type promotion will happen implicitly also for `p > 0`, it's just more explicit. It also seems to work fine for `Dual`s:
```julia
julia> using ForwardDiff

julia> oftype(ForwardDiff.Dual(300.0), -Inf)
Dual{Nothing}(-Inf)

julia> typeof(oftype(ForwardDiff.Dual(300.0), -Inf))
ForwardDiff.Dual{Nothing, Float64, 0}

julia> typeof(oftype(ForwardDiff.Dual(300f0), -Inf))
ForwardDiff.Dual{Nothing, Float32, 0}

julia> oftype(ForwardDiff.Dual(300f0), -Inf)
Dual{Nothing}(-Inf)
```